storm-core/src/jvm/org/apache/storm/daemon/supervisor/Slot.java

```diff
@@ -0,0 +1,769 @@
+
+                if (LOG.isDebugEnabled() || dynamicState.state != nextState.state) {
+                    LOG.info("STATE {} -> {}", dynamicState, nextState);
+                }
+                //Save the current state for recovery
+                if (!equivilant(nextState.currentAssignment, dynamicState.currentAssignment)) {
+                    LOG.warn("SLOT {}: Changing current assignment from {} to {}", staticState.port, dynamicState.currentAssignment, nextState.currentAssignment);
+                    synchronized(staticState.localState) {
+                        Map<Integer, LocalAssignment> assignments = staticState.localState.getLocalAssignmentsMap();
+                        if (assignments == null) {
+                            assignments = new HashMap<>();
+                        }
+                        if (nextState.currentAssignment == null) {
+                            assignments.remove(staticState.port);
+                        } else {
+                            assignments.put(staticState.port, nextState.currentAssignment);
+                        }
+                        staticState.localState.setLocalAssignmentsMap(assignments);
+                    }
+                    Map<Long, LocalAssignment> update = null;
+                    Map<Long, LocalAssignment> orig = null;
+                    do {
+                        Long lport = new Long(staticState.port);
+                        orig = cachedCurrentAssignmants.get();
+                        update = new HashMap<>(orig);
+                        if (nextState.currentAssignment == null) {
+                            update.remove(lport);
+                        } else {
+                            update.put(lport, nextState.currentAssignment);
+                        }
+                    } while (!cachedCurrentAssignmants.compareAndSet(orig, update));
+                }
+                
+                // clean up the profiler actions that are not being processed
+                removed.removeAll(dynamicState.profileActions);
+                removed.removeAll(dynamicState.pendingStopProfileActions);
+                for (TopoProfileAction action: removed) {
+                    try {
+                        clusterState.deleteTopologyProfileRequests(action.topoId, action.request);
+                    } catch (Exception e) {
+                        LOG.error("Error trying to remove profiling request, it will be retried", e);
+                    }
+                }
+                Set<TopoProfileAction> orig, copy;
+                do {
+                    orig = profiling.get();
+                    copy = new HashSet<>(orig);
+                    copy.removeAll(removed);
+                } while (!profiling.compareAndSet(orig, copy));
+                dynamicState = nextState;
+                newAssignment.compareAndSet(localNewAssignment, dynamicState.newAssignment);
```

Comment:
Ok. Thanks! Took me a while to go through Slot and the state machine. The unit tests were very helpful on that.

To summarize how I understand this:
1. we get a copy of new assignment
2. mutate localstate assignments
3. mutate supervisor in-memory assingments
4. deal with profiling cleanup
5. persist the new assignment in slot.
6. If new assignment persisted successfully, then we are good. The equivalency checks will pass, and the state machine should progress or stay at RUNNING. Else, the equivalency checks will move us to correct state. Also this loop fixes up localstate and in-memory assignments continuously.